### PR TITLE
qt: fix for ARM

### PR DESCRIPTION
**Pull request by fxcoudert**
Allow Qt to compile on ARM by curtailing its features a bit (see https://github.com/Homebrew/brew/issues/7857#issuecomment-747540065)

The reason is that a backport of the upstream webengine fixes (made to the Qt 6.0 branch) would be huge.

### CODE REVIEW

**Comment by Mushroom** [comment]
Thanks for opening this and doing it the proper way, as opposed to the slight hack I did.

As per the reply [here](https://github.com/Homebrew/brew/issues/7857#issuecomment-748170325) we might need to patch `qtbase/mkspecs` also, to prevent projects built on M1 machines from generating x64 binaries by default

**Comment by fxcoudert** [comment]
@Mushroom can you point to the upstream patch for qtbase/mkspecs please?

**Comment by Mushroom** [comment]
Looks like people are having the issue upstream also (look at the most recent comments on [this](https://bugreports.qt.io/browse/QTBUG-85279)).

I can look into what needs to be done to fix this tomorrow, if required.

**Comment by Mushroom** [comment]
@fxcoudert I have created a commit here that fixes this issue: https://github.com/Mushroom/qtbase/commit/abc52460201bc5c7603505bb187138b0c59291aa

Should I try and push this up to the Qt project first, or shall it just be cherry picked and we wait for 6.0 to sort things properly, much like with the web engine?

**Comment by fxcoudert** [comment]
> Should I try and push this up to the Qt project first

Please see what they think about it, even to cherry-pick it we prefer that it be accepted or validated by upstream in some way

**Comment by Mushroom** [comment]
Done: https://codereview.qt-project.org/c/qt/qtbase/+/327649

Just need to wait for their feedback now

**Comment by torarnv** [comment]
Mushroom/qtbase@abc5246 looks good as a workaround for now. Thanks @Mushroom! 😊

**Comment by Mushroom** [comment]
@torarnv No problem, thanks for the prompt review!

@fxcoudert Is there anything you need from my end for this now?

**Comment by MaiReo** [comment]
After applying this patch locally (brew edit qt, copy and paste)
brew install qt shows
Error: No available formula or cask with the name "qt".

**Comment by SMillerDev** [comment]
That means it can't read the formula because you introduced a syntax issue.

**Comment by fxcoudert** [comment]
Updating with patch (thanks @Mushroom @torarnv), running tests on ARM.

**Comment by fxcoudert** [comment]
@Mushroom the patch is leading to build failure:

```
2020-12-24T12:46:33.7656980Z c -o main.o -fconstant-cfstrings -MMD     -g  -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/qmake -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/qmake/library -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/qmake/generators -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/src/3rdparty/tinycbor/src -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/qmake/generators/unix -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/qmake/generators/win32 -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/qmake/generators/mac -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/include -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/include/QtCore -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/include/QtCore/5.15.2 -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/include/QtCore/5.15.2/QtCore -I../src/corelib/global -I/private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/mkspecs/macx-clang-arm64 -DQT_VERSION_STR=\"5.15.2\" -DQT_VERSION_MAJOR=5 -DQT_VERSION_MINOR=15 -DQT_VERSION_PATCH=2 -DQT_BUILD_QMAKE -DQT_BOOTSTRAPPED -DPROEVALUATOR_FULL -DQT_NO_FOREACH /private/tmp/qt-20201224-49924-113yvid/qt-everywhere-src-5.15.2/qtbase/qmake/main.cpp
2020-12-24T12:46:33.7665280Z make: c: No such file or directory
```

and

```
2020-12-24T12:46:33.9669060Z o "../bin/qmake" main.o meta.o option.o project.o property.o ioutils.o proitems.o qmakebuiltins.o qmakeevaluator.o qmakeglobals.o qmakeparser.o qmakevfs.o pbuilder_pbx.o makefile.o makefiledeps.o metamakefile.o projectgenerator.o unixmake2.o unixmake.o mingw_make.o msbuild_objectmodel.o msvc_nmake.o msvc_objectmodel.o msvc_vcproj.o msvc_vcxproj.o winmakefile.o xmloutput.o qutfcodec.o qendian.o qglobal.o qlogging.o qmalloc.o qnumeric.o qoperatingsystemversion.o qrandom.o qabstractfileengine.o qbuffer.o qdatastream.o qdebug.o qdir.o qdiriterator.o qfile.o qfiledevice.o qfileinfo.o qfilesystemengine.o qfilesystementry.o qfsfileengine.o qfsfileengine_iterator.o qiodevice.o qsettings.o qtemporaryfile.o qtextstream.o qcborstreamwriter.o qcborvalue.o qjsoncbor.o qjsonarray.o qjsondocument.o qjsonobject.o qjsonparser.o qjsonvalue.o qmetatype.o qsystemerror.o qvariant.o quuid.o qarraydata.o qbitarray.o qbytearray.o qbytearraylist.o qbytearraymatcher.o qcalendar.o qgregoriancalendar.o qromancalendar.o qcryptographichash.o qdatetime.o qhash.o qlist.o qlocale.o qlocale_tools.o qmap.o qregexp.o qringbuffer.o qstringbuilder.o qstring.o qstringlist.o qversionnumber.o qvsnprintf.o qxmlstream.o qxmlutils.o qcore_unix.o qfilesystemengine_unix.o qfilesystemiterator_unix.o qfsfileengine_unix.o qlocale_unix.o qsettings_mac.o qcore_mac.o qoperatingsystemversion_darwin.o qcore_foundation.o qlibraryinfo.o -framework Foundation -framework CoreServices -framework ApplicationServices  
2020-12-24T12:46:33.9678850Z /bin/sh: o: command not found
```

I'm not sure where `c` and `o` are coming from…

see full log at https://pipelines.actions.githubusercontent.com/revSbsLpzrFApLL6BmCvScWt72tRe3wYUv7fCdCtThtI5bydk7/_apis/pipelines/1/runs/103235/signedlogcontent/17?urlExpires=2020-12-24T12%3A51%3A47.7619019Z&urlSigningMethod=HMACV1&urlSignature=0dI6fHW4S3Lx%2BNpe8C%2FBe5dsQYbOarH7GJo61l01F%2Fo%3D

**Comment by Mushroom** [comment]
I'll have a look - never seen that before though

**Comment by Mushroom** [comment]
Looks like it's picking up "c" instead of "clang", when I compare it to my build - no idea why it would do that

**Comment by fxcoudert** [comment]
I don't know either, all I do is apply the patch directly 😢

**Comment by Mushroom** [comment]
Does it work locally for you? Might be an env issue in the CI

**Comment by fxcoudert** [comment]
> Does it work locally for you?

I don't own an M1 machine :) I'm just trying to help out but putting things together, and testing on our CI.

**Comment by Mushroom** [comment]
Makes sense - let's see if someone else with an M1 mac can reproduce it locally then, and if so, I'll see what's different on my side that allows it to work.

**Comment by fxcoudert** [comment]
@Mushroom can you apply the homebrew patch as it is in this PR, and confirm that it works for you? That would be a start.

File: Formula/qt.rb [hunk 1]
```diff
@@ -56,9 +66,16 @@ def install
...
       -pkg-config
       -dbus-runtime
-      -proprietary-codecs
     ]
 
+    if Hardware::CPU.arch == :arm64
+      # Temporarily fixes for Apple Silicon
+      args << "-skip" << "qtwebengine" << "-no-assimp" << "QMAKE_APPLE_DEVICE_ARCHS=arm64"
+    else
+      # Should be reenabled unconditionnaly once it is fixed on Apple Silicon
```
Comment (MikeMcQuaid) [inline Formula/qt.rb:75]:
```suggestion
      # Should be reenabled unconditionally once it is fixed on Apple Silicon
```

File: Formula/qt.rb [hunk 2]
```diff
@@ -56,9 +66,16 @@ def install
       -no-rpath
       -pkg-config
       -dbus-runtime
-      -proprietary-codecs
     ]
 
+    if Hardware::CPU.arch == :arm64
+      # Temporarily fixes for Apple Silicon
+      args << "-skip" << "qtwebengine" << "-no-assimp" << "QMAKE_APPLE_DEVICE_ARCHS=arm64"
```
Comment (MikeMcQuaid) [inline Formula/qt.rb:73]: May want to note the `qtwebengine` omission in `caveats`?

Comment (fxcoudert) [reply Formula/qt.rb:73]: Yeah, if we get it working, I'll add some caveats

**Comment by Mushroom** [comment]
I just hit the same issue with your patch locally.

When I went back to master, and just added the changes from my comment, it did work.

**Comment by Mushroom** [comment]
Had a further dig into this - it looks like the patch isn't being applied properly - the adds are going through okay, but the moves are not being applied by the looks of it, so `macx-clang/qmake.comnf` is not being moved to `mkspecs/common/clang-macx-desktop.conf`

I don't know much about the patch mechanism in homebrew, perhaps there is a bug in it related to renames?

**Comment by fxcoudert** [comment]
We use `patch` to apply patches, more specifically, `patch -g 0 -f -p1 -i abc5246.patch`
I'm not sure why this is not working with renamed files…

**Comment by Mushroom** [comment]
It's possibly reproducible on an x64 mac, if you have one of those. This patch should affect both architectures as it splits them.

**Comment by Mushroom** [comment]
Ah - is there anything that can be done to work around it? I could see if I can generate a non-git style patch and host it somewhere

**Comment by fxcoudert** [comment]
I've generated a diff with `--no-renames` from your git tree and will host it.

**Comment by fxcoudert** [comment]
And we try again… thanks for investigating the root cause.

**Comment by Mushroom** [comment]
No worries - we can get rid of the `QMAKE_APPLE_DEVICE_ARCHS=arm64` in the args with that patch also, but it will work with it also without issue

**Comment by fxcoudert** [comment]
```
In file included from main.cpp:1:
In file included from /opt/homebrew/Cellar/qt/5.15.2/lib/QtCore.framework/Headers/QCoreApplication:1:
In file included from /opt/homebrew/Cellar/qt/5.15.2/lib/QtCore.framework/Headers/qcoreapplication.h:46:
In file included from /opt/homebrew/Cellar/qt/5.15.2/lib/QtCore.framework/Headers/qobject.h:49:
/opt/homebrew/Library/Homebrew/shims/scm/git --version
In file included from /opt/homebrew/Cellar/qt/5.15.2/lib/QtCore.framework/Headers/qlist.h:47:
In file included from /opt/homebrew/Cellar/qt/5.15.2/lib/QtCore.framework/Headers/qhashfunctions.h:47:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/numeric:148:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/cmath:317:9: error: no member named 'signbit' in the global namespace
using ::signbit;
      ~~^
```

**Comment by Mushroom** [comment]
What version of the SDK is installed? I tested this against 11.1

**Comment by fxcoudert** [comment]
We have macOS 11.1 with Xcode 12.3.
On occasions in the past this issue was linked to `SDKROOT` being set, I don't have time to debug right now but I can try unsetting it 🤞

**Comment by Mushroom** [comment]
Was there any luck with this @fxcoudert? As I can't reproduce this locally I can't really help

**Comment by fxcoudert** [comment]
@Mushroom no that didn't work

**Comment by Mushroom** [comment]
Do you have a script of some sort to reproduce the CI env exactly? I'll try with a VM or Container of some sort to reproduce the issue.

**Comment by fxcoudert** [comment]
`brew install -svd qt` after having applied to patch in this PR

**Comment by Mushroom** [comment]
That's pretty much what I'm doing - do you have a script you used to set up the CI node or was it done by hand?

**Comment by jonaski** [comment]
Try: `system("make", "CPATH=")`

**Comment by fxcoudert** [comment]
@Mushroom the only difference I can think of is: do you have CLT, Xcode, or both? Our CI has both.

**Comment by Mushroom** [comment]
@fxcoudert Both. Could you pastebin a printenv from the CI machine (redacted if necessary)

**Comment by fxcoudert** [comment]
@Mushroom what are you expecting to find? the machine is basically a clean macOS, except Xcode, CLT, and standard brew install:

```
LC_ALL=en_US.UTF-8
LANG=fr_FR.UTF-8
USER=me
LOGNAME=me
HOME=/Users/me
PATH=/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
SHELL=/bin/zsh
TERM=xterm
SSH_CLIENT=redacted
SSH_CONNECTION=redacted
SSH_TTY=/dev/ttys000
TMPDIR=/var/folders/vc/9n6c_r253cddwcx2lgcqd1280000gp/T/
SHLVL=1
PWD=/Users/me
OLDPWD=/Users/me
_=/usr/bin/printenv
```

**Comment by Mushroom** [comment]
@fxcoudert Just to compare it to mine locally for even the smallest difference! I'm having a hack about with parallels right now to see if I can get a clean VM working, and to see if I can reproduce.

**Comment by fxcoudert** [comment]
@jonaski thanks, `unset CPATH` actually makes the test pass. I'd like to understand why, though, before we proceed with merging this pull request. Is it working around a bug, or why does this happen at all?

**Comment by fxcoudert** [comment]
Passes on ARM, no-op elsewhere (and tested locally on Intel): merging.

**Comment by fxcoudert** [comment]
@Mushroom now `qwt` fails to build on ARM. It calls `qmake -config release -spec macx-clang` and complains about `Could not find qmake spec 'macx-clang'`. What would be the substitute in that case?

**Comment by Mushroom** [comment]
@fxcoudert They would be:
* `macx-clang-arm64` for `aarch64`
* `macx-clang-x64` for `x86_64`

**Comment by fxcoudert** [comment]
@Mushroom but that's only for your patch, right? Do you know how upstream Qt intends to fix things in Qt 6?

**Comment by Mushroom** [comment]
That's a question for @torarnv I think. For now just removing that argument should make it build the same as the host architecture/OS.

**Comment by torarnv** [comment]
As long as you're using a separate makespec you don't need to pass the QMAKE_APPLE_DEVICE_ARCHS option to configure.

**Comment by ticky** [comment]
Should this change have bumped the formula's revision? I just had to manually uninstall and reinstall to get these changes. Otherwise, this is great, thanks for the fix :)

**Comment by fxcoudert** [comment]
Well ARM was not really supported before, hence no revision.

**Pull request merged**